### PR TITLE
CORE-14178: merge forward 5.0 changes to 5.1

### PR DESCRIPTION
**Pull request by yift-r3**
API pull request in https://github.com/corda/corda-api/pull/1141

Contains:
* https://github.com/corda/corda-runtime-os/pull/3980
* https://github.com/corda/corda-runtime-os/pull/4009
* https://github.com/corda/corda-runtime-os/pull/3976
* https://github.com/corda/corda-runtime-os/pull/3998
* https://github.com/corda/corda-runtime-os/pull/3960
* https://github.com/corda/corda-runtime-os/pull/3990
* https://github.com/corda/corda-runtime-os/pull/3993
* https://github.com/corda/corda-runtime-os/pull/3964
* https://github.com/corda/corda-runtime-os/pull/3991
* https://github.com/corda/corda-runtime-os/pull/3982
* https://github.com/corda/corda-runtime-os/pull/3981
* https://github.com/corda/corda-runtime-os/pull/3959
* https://github.com/corda/corda-runtime-os/pull/3983
* https://github.com/corda/corda-runtime-os/pull/3942
* https://github.com/corda/corda-runtime-os/pull/3956
* https://github.com/corda/corda-runtime-os/pull/3962
* https://github.com/corda/corda-runtime-os/pull/3957
* https://github.com/corda/corda-runtime-os/pull/3971
* https://github.com/corda/corda-runtime-os/pull/4003

### CODE REVIEW

**Comment by chrisr3** [comment]
This looks like a repeat of #4007.

**Comment by dickon** [comment]
Title should be CORE-13934 not CORE-0, as per https://r3-cev.atlassian.net/wiki/spaces/CB/pages/4396089494/Keeping+release+branches+in+Sync

**Comment by yift-r3** [comment]
> Title should be CORE-13934 not CORE-0, as per https://r3-cev.atlassian.net/wiki/spaces/CB/pages/4396089494/Keeping+release+branches+in+Sync

Why CORE-13934?!

**Review by dickon — DISMISSED** [review]
LGTM for crypto changes.

I also checked both the changes to crypto code from @chrisr3  and @conalsmith-r3 look good. I didn't look at the non-crypto parts of @chrisr3's changes.

File: gradle.properties [hunk 1]
```diff
@@ -44,7 +44,7 @@ bouncycastleVersion=1.73
 # Corda API libs revision (change in 4th digit indicates a breaking change)
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
 #cordaApiVersion=5.0.0.xxx-SNAPSHOT
-cordaApiVersion=5.1.0.4-beta+
+cordaApiVersion=5.1.0.5-alpha-1685952683252
```
Comment (vkolomeyko) [inline gradle.properties]: We would need to change this to Beta version once `corda-api` PR is merged.

**Review by vkolomeyko — DISMISSED** [review]

**Review by williamvigorr3 — DISMISSED** [review]
LGTM (for my commit)

**Review by Tom-Fitzpatrick — DISMISSED** [review]
LGTM for my commit

**Review by nkovacsx — DISMISSED** [review]
LGTM (my changes)

**Review by chrisr3 — DISMISSED** [review]
What you have of my changes look OK, but I am surprised that this merge - which was only created today - doesn't also include
```
Author: Chris Rankin <chris.rankin@r3.com>
Date:   Sat Jun 3 16:19:05 2023 +0100

    CORE-14320: Create a ComputedValue gauge metric. (#4003)
```

**Review by dimosr — DISMISSED** [review]
LGTM (for my commits)

**Review by chrisr3 — DISMISSED** [review]
LGTM

**Review by jujoramos — DISMISSED** [review]
LGTM

**Review by williamvigorr3 — DISMISSED** [review]
LGTM (my commit).

**Review by charlieR3 — DISMISSED** [review]
The changes from the following PRs LGTM:
https://github.com/corda/corda-runtime-os/pull/3993
https://github.com/corda/corda-runtime-os/pull/3942

**Review by fowlerrr — DISMISSED** [review]
My change looks good

**Review by vkolomeyko — DISMISSED** [review]

**Review by JamesHR3 — DISMISSED** [review]

**Review by conal-smith — DISMISSED** [review]

**Review by vkolomeyko — APPROVED** [review]
LGTM

**Review by dimosr — APPROVED** [review]

**Review by wzur-r3 — APPROVED** [review]
LGTM for Gradle files

**Review by fowlerrr — APPROVED** [review]

**Comment by dickon** [comment]
> Why CORE-13934?!

first ticket of the commits that are being merge across. I don't suppose it really matters which one you take, or if it is a new ticket for doing the merge,  but ideally we'd be doing this on every commit so it'd be 1:1. However, CORE-0 gives us no information and would get flagged in source code accountability audits.

**Pull request merged**